### PR TITLE
PowerShell - Added coverage collector

### PR DESCRIPTION
**Pull request by jborean93**
##### SUMMARY
Added a wrapper that will collect code coverage of powershell modules and module_utils.

TODO:

* ~Integration into ansible-test to set the `_ANSIBLE_COVERAGE_OUTPUT` env var and retrieve the coverage files once complete~
* ~Look into improving performance in any way possible.~
* ~Implement a way to filter the files we collect coverage when testing a collection.~ - Done, only files in the current content will be reported on

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ansible-test
powershell

### CODE REVIEW

**Comment by jborean93** [comment]
I've had a look into the performance problems and have been able to improve the time it takes for a module to run. Still not the same as a non coverage run but the majority of the extra time is now taken up by internal functions we have no control over.

I have also looked into [Set-PSDebug](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/set-psdebug?view=powershell-6) to see if line tracing is a better candidate for coverage collection but it is not feasible because;

* The scope of the trace levels are set on a script level, we cannot set it to the pipeline/runspace globally and therefore would have to edit the module code to enable it
* The output is just a string, it is easily parsible but extra work and not sure if the format is guaranteed in future versions
* I could not find a way to collect coverage for PowerShell module util code (.psm1) so this would only be good for our Windows modules
* After tweaking the performance a bit for the breakpoint method, the biggest obstacle is no longer there

Now that I am somewhat happy with the actual coverage collection process, I need to figure out the best way to integrate that into ansible-test and generate the code coverage xml report to be sent to Code Cov. The following questions need to be answered;

* How can we integrate the PowerShell coverage info into the xml generation process
* Can we somehow report on line branches coverage (if statement evaluation)
* Should we even use the .xml or just use the basic [CodeCov .json](https://docs.codecov.io/reference#section-codecov-json-report-format) format
* When fetching the coverage results back should that be merged into the 1 file or kept separate like we do with `test/reulsts/coverage/*`

**Review by nitzmahone — COMMENTED** [review]
Looks great! I'm not that familiar with the overall ansible-test integration coverage mechanism- does it have a way to collect the files from the targets? IIRC the current Python stuff largely assumes that it's running on the controller... I know Matt wanted to work on integrating the coverage results more deeply (eg maybe smuggling them back directly in the task results), but I don't think any of that exists yet. I guess the question from all of this is "what (if anything) is missing to bring this to parity with the existing Python coverage stuff?" (PS, I always forget about draft PRs, but perfect case for one!)

**Comment by jborean93** [comment]
> does it have a way to collect the files from the targets

Not yet, as you said it relies on it being on the same host so part of this PR that I need to modify is to retrieve the results back at the end of the run.

> I know Matt wanted to work on integrating the coverage results more deeply (eg maybe smuggling them back directly in the task results), but I don't think any of that exists yet

I briefly looked at this but it would involve us either;

* Reparsing the output json, adding the results then outputting the json again with the results with the controller doing what it needs to
* Somehow get the controller to accept 2 json data blocks, 1 with and without the coverage info

Both mechanisms would mean we are sending more data back to the controller as well as formatting the results in each task which will add even more time per task execution. The route that I am planning means the biggest overhead is the setting of the breakpoints and running with coverage so it's as slim as I can get it. The result processing and retrieving the data happens in 1 big go right at the end but I still need to implement that.

> I guess the question from all of this is "what (if anything) is missing to bring this to parity with the existing Python coverage stuff?

The main hurdle that is currently "blocking" me is how to reconcile the coverage data from this to the Python work. Python uses a branching coverage report but PowerShell cannot do this so there's no way to integrate the data into the 1 report easily. Python coverage.py does not allow you to do this so I'm either going to have to just submit 2 reports (1 with line based coverage and 1 with branch based coverage) or somehow integrate the line based work into the 1 xml repo. I'm leaning towards the former as I don't particular want to process XML files but have yet to start on that work.

File: test/lib/ansible_test/_internal/executor.py [hunk 1]
```diff
@@ -658,16 +660,43 @@ def cleanup_ssh_ports(target):
             pre_target = forward_ssh_ports
             post_target = cleanup_ssh_ports
 
+    def run_playbook(playbook, playbook_vars):
```
Comment (mattclay) [inline test/lib/ansible_test/_internal/executor.py:684]: This looks like something that would have more general use in the future and should probably go in `util.py` instead. Thoughts?

File: test/lib/ansible_test/_internal/executor.py [hunk 2]
```diff
@@ -658,16 +660,43 @@ def cleanup_ssh_ports(target):
...
 
+        if remote_temp_path:
+            # Zip up the coverage files that were generated and fetch it back to localhost.
+            with tempdir() as local_temp_path:
+                playbook_vars = {'remote_temp_path': remote_temp_path, 'local_temp_path': local_temp_path}
+                run_playbook('windows_coverage_teardown.yml', playbook_vars)
+
+                for filename in os.listdir(local_temp_path):
+                    with open_zipfile(os.path.join(local_temp_path, filename)) as coverage_zip:
+                        coverage_zip.extractall(os.path.join(data_context().results, 'coverage'))
```
Comment (mattclay) [inline test/lib/ansible_test/_internal/executor.py:719]: Can we use `ZipFile.extract` instead of `ZipFile.extractall` here?

Comment (jborean93) [reply test/lib/ansible_test/_internal/executor.py:719]: Why is that, it's a lot simpler to use extractall compared to iterating the zip contents and then extracting each member.

File: test/lib/ansible_test/_internal/executor.py (hunk 1 above)
Comment (jborean93) [reply test/lib/ansible_test/_internal/executor.py:684]: I didn't at the time as it has a few things specific for the coverage setup/destroy playbooks, namely

* The inventory is set outside
* We are running without coverage being enabled
* Not sure if the playbook dir would ever need to be configured but it was set to the 1 here for simplicity

We can move it I just didn't want to introduce common code without thinking of a stablish interface to expose.

File: test/lib/ansible_test/_internal/executor.py (hunk 2 above)
Comment (mattclay) [reply test/lib/ansible_test/_internal/executor.py:719]: Does member have to be a file? Or can it be a subdirectory?

File: test/lib/ansible_test/_internal/executor.py (hunk 1 above)
Comment (mattclay) [reply test/lib/ansible_test/_internal/executor.py:684]: Leaving it here is fine. I was just curious what your thoughts were.

File: test/lib/ansible_test/_internal/executor.py (hunk 2 above)
Comment (jborean93) [reply test/lib/ansible_test/_internal/executor.py:719]: You can do both, the biggest advantage `.extract()` offers is the ability to filter what files to extract whereas `.extractall()` does them all. For the PowerShell coverage we just put the files in the root directory.

**Review by mattclay — APPROVED** [review]

**Comment by jborean93** [comment]
While we do have a timeout we need to solve this is an ongoing issue with coverage runs, merging to get this in before the freeze and to enable other development.

**Pull request merged**